pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -1084,7 +1084,8 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 
 		switch status.State {
 		case kubecontainer.ContainerStateCreated:
-			// nothing to do but wait for it to start
+			// restart the container if it is stuck in the created state
+			changes.InitContainersToStart = append(changes.InitContainersToStart, i)
```

Comment:
Can you explain why container considered stuck if it's in the `ContainerStateCreated` state?

Reply:
That is because of the old code that does nothing in the Created state.
I'll rephrase it.

Reply:
It still looks to me like we're trying to unconditionally restart container if it's in the `ContainerStateCreated` state. Where this state is set? Where is it updated? Why is it `likely that the container runtime has failed to start the container` if the container is in this state?

/hold

Reply:
This is a part of the main sync loop and
1. it checks the state
2. calls the CreateContainer
3. calls the StartContainer

in sequence with blocking manner (waiting for the result).

Therefore, if a container is in the Created state at step 1, it means that the container runtime has failed to StartContainer.

https://github.com/kubernetes/kubernetes/pull/126543/files#r1704785057
^
The regular containers behave in the same way.

Reply:
It does. Thanks for the explanations!
/unhold

Reply:
This can cause the same issue as #123087 for init containers when `EventedPLEG` is enabled.

Container statuses evaluated here are updated asynchronously by PLEG. So, `ContainerStateCreated` could be seen when a container is starting normally especially with `EventedPLEG` enabled.
xref: https://github.com/kubernetes/kubernetes/pull/122778#issuecomment-1956910606

Reply:
Thanks for sharing! I see that the change to prevent Regular containers from restarting in the Created state when Evented PLEG is enabled will be made in https://github.com/kubernetes/kubernetes/pull/122778.

This PR fixes a regression in Generic PLEG caused by Sidecar Containers feature. We could either merge this PR first and then address the case for initContainers in https://github.com/kubernetes/kubernetes/pull/122778, or we could take a different approach. Based on the  https://github.com/kubernetes/enhancements/issues/3386#issuecomment-2309247626, it looks like there might be some progress with https://github.com/kubernetes/kubernetes/pull/122778 in near future too.

/cc @pacoxu

Reply:
This addressed the problem of Generic PLEG. 

/lgtm
/approve

Reply:
@gjkim42 if (and only if) you happen to resubmit, could you please add more of your comment https://github.com/kubernetes/kubernetes/pull/126543/commits/7e8ea57cc6ee29631e8a66cc521a48803a16cd14#r1724882374 in the code comment? I feel the github comment concisely and clearly describes the fix so it would be great to have it in the codebase and not lost in github review comments.

Reply:
resubmitted.